x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStoreCacheTests.java

```diff
@@ -0,0 +1,303 @@
+    private static final String APP_USER_NAME = "app_user";
+
+    @Override
+    protected String configUsers() {
+        return super.configUsers()
+            + APP_USER_NAME + ":" + TEST_PASSWORD_HASHED + "\n";
+    }
+
+    @Override
+    protected String configRoles() {
+        return super.configRoles()
+            + "app_role:\n"
+            + "  cluster: ['monitor']\n"
+            + "  indices:\n"
+            + "    - names: ['*']\n"
+            + "      privileges: ['read']\n"
+            + "  applications:\n"
+            + "    - application: 'app-1'\n"
+            + "      privileges: ['read', 'check']\n"
+            + "      resources: ['foo']\n"
+            + "    - application: 'app-2'\n"
+            + "      privileges: ['check']\n"
+            + "      resources: ['foo']\n";
+    }
+
+    @Override
+    protected String configUsersRoles() {
+        return super.configUsersRoles()
+            + "app_role:" + APP_USER_NAME + "\n"
+            + TEST_ROLE + ":" + APP_USER_NAME + "\n";
+    }
+
+    @Before
+    public void configureApplicationPrivileges() {
+        final List<ApplicationPrivilegeDescriptor> applicationPrivilegeDescriptors = Arrays.asList(
+            new ApplicationPrivilegeDescriptor("app-1", "read", Set.of("a:b:c", "x:y:z"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-1", "write", Set.of("a:b:c", "x:y:z"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-1", "admin", Set.of("a:b:c", "x:y:z"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-2", "read", Set.of("e:f:g", "t:u:v"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-2", "write", Set.of("e:f:g", "t:u:v"), emptyMap()),
+            new ApplicationPrivilegeDescriptor("app-2", "admin", Set.of("e:f:g", "t:u:v"), emptyMap()));
+
+        final PutPrivilegesRequest putPrivilegesRequest = new PutPrivilegesRequest();
+        putPrivilegesRequest.setPrivileges(applicationPrivilegeDescriptors);
+        final ActionFuture<PutPrivilegesResponse> future =
+            client().execute(PutPrivilegesAction.INSTANCE, putPrivilegesRequest);
+
+        final PutPrivilegesResponse putPrivilegesResponse = future.actionGet();
+        assertEquals(2, putPrivilegesResponse.created().size());
+        assertEquals(6, putPrivilegesResponse.created().values().stream().mapToInt(List::size).sum());
+    }
+
+    public void testGetPrivileges() {
+        final Client client = client();
+
+        ApplicationPrivilegeDescriptor[] privileges = new GetPrivilegesRequestBuilder(client)
+            .application("app-2").privileges("write").execute().actionGet().privileges();
+
+        assertEquals(1, privileges.length);
+        assertEquals("app-2", privileges[0].getApplication());
+        assertEquals("write", privileges[0].getName());
+
+        // A hacky way to test cache is populated and used by deleting the backing documents.
+        // The test will fail if the cache is not in place
```

Comment:
I think this is a sign that this test shouldn't be an integration test like this. If it weren't quite so integrated you would be able to spy on the client used by the privilege store.